### PR TITLE
Add sequence number to bookmark and simplify URIGraph updating

### PR DESCRIPTION
**Pull request by gmethvin**
Use a sequence number in the bookmark table to determine which users
to update both on load of the graph and on update. Remove the
URIGraph#update(user) method and use a single update() method for
loading and updating the graph.

### CODE REVIEW

File: shoebox/app/com/keepit/search/graph/URIGraph.scala [hunk 1]
```diff
@@ -76,47 +73,37 @@ class URIGraphImpl(indexDirectory: Directory, indexWriterConfig: IndexWriterConf
...
-  def update(userId: Id[User]): Int = {
-    log.info("updating a URIGraph for user=%d".format(userId.id))
-    try {
-      val user = inject[Database].readOnly { implicit s => inject[UserRepo].get(userId) }
-      var cnt = 0
-      indexDocuments(Iterator(buildIndexable(user)), commitBatchSize){ commitBatch =>
+      val usersChanged = inject[Database].readOnly { implicit s =>
+        inject[BookmarkRepo].getUsersChanged(sequenceNumber)
+      }
+      val indexables = usersChanged.map(buildIndexable(_)).toSeq
```
Comment (andrewconner) [inline shoebox/app/com/keepit/search/graph/URIGraph.scala]: If you're doing shorthand, could be `usersChanged.map(buildIndexable)`. `Set` has a `.iterable` as well, so `.toSeq` isn't needed.

File: shoebox/app/com/keepit/search/graph/URIGraphPlugin.scala [hunk 2]
```diff
@@ -70,13 +61,8 @@ class URIGraphPluginImpl @Inject() (system: ActorSystem, uriGraph: URIGraph) ext
...
-  override def load(): Int = {
-    val future = actor.ask(Load)(1 minutes).mapTo[Int]
-    Await.result(future, 1 minutes)
-  }
-
-  override def update(userId: Id[User]) = {
-    val future = actor.ask(Update(userId))(1 minutes).mapTo[Int]
+  override def update() = {
+    val future = actor.ask(Update)(1 minutes).mapTo[Int]
     Await.result(future, 1 minutes)
```
Comment (andrewconner) [inline shoebox/app/com/keepit/search/graph/URIGraphPlugin.scala]: Is it possible for this to return a `Future` instead? Typically, Akka docs discourage using `Await`, and if you must, to do it only at the outside layers.

File: shoebox/app/com/keepit/search/index/Indexer.scala [hunk 3]
```diff
@@ -107,26 +107,26 @@ abstract class Indexer[T](indexDirectory: Directory, indexWriterConfig: IndexWri
...
     if (refresh) refreshSearcher()
   }
 
-  def commit() {
-    indexWriter.commit(Map(
-      Indexer.CommitData.committedAt -> currentDateTime.toStandardTimeString,
-      Indexer.CommitData.sequenceNumber -> sequenceNumber.toString
+  def commit(iw: IndexWriter) {
+    iw.commit(Map(
+          Indexer.CommitData.committedAt -> currentDateTime.toStandardTimeString,
```
Comment (andrewconner) [inline shoebox/app/com/keepit/search/index/Indexer.scala]: Not a knock on this, but as a group, we should probably standardize on a way to get the current time. We moved to `inject[DateTime]`, but I know you had some reservations about that. Maybe we should quickly discuss which one we should use.

**Comment by andrewconner** [comment]
I like this, especially how the indexer now rebuilds for all updated users. Should we have the functionality to specifically do one, or do you think that this will cover us?

File: shoebox/app/com/keepit/search/index/Indexer.scala (hunk 3 above)
Comment (gmethvin) [inline shoebox/app/com/keepit/search/index/Indexer.scala]: I think the proper Guice way to do this is to inject a `Provider[DateTime]` into the class. Maybe we could make `currentDateTime` require an implicit provider in scope.

I would prefer that whatever solution we have rely on global state as little as possible.

**Comment by gmethvin** [comment]
@andrewconner We could have the functionality to specifically do one user, but I don't see a use case for it that isn't covered by updating the changed records.

**Comment by gmethvin** [comment]
@ymatsuda any comments on this?

**Comment by ymatsuda** [comment]
I think we need to have a way to force reindexing of a specific user via an admin interface. It doesn't necessarily mean we need to have a method for that. The admin interface can set a new sequence id of a users bookmark. This is a convenient way to force all search servers to update indexes. Even when a server is down at that moment, it will reindex the user after it starts up.

**Comment by gmethvin** [comment]
@ymatsuda I see. The easiest way would be to update the sequence numbers of all the bookmarks of that user, then call URIGraph#update(). I can do that for the update(user) controller call.

File: shoebox/app/com/keepit/search/index/Indexer.scala (hunk 3 above)
Comment (ymatsuda) [inline shoebox/app/com/keepit/search/index/Indexer.scala]: why is an indexwriter an argument of the commit method? Indexer is supposed to encapsulate indexwriter. I may be missing something, but passing it as an argument doesn't seem right.

Comment (eishay) [inline shoebox/app/com/keepit/search/index/Indexer.scala]: `currentDateTime` is discouraged. As we'll do more guicing we should
inject datatime or a clock.

**Comment by gmethvin** [comment]
Now clicking update on the bookmarks section of the user admin page should force a refresh of the user's URIGraph.

File: shoebox/app/com/keepit/search/index/Indexer.scala (hunk 3 above)
Comment (gmethvin) [inline shoebox/app/com/keepit/search/index/Indexer.scala]: @ymatsuda I required the indexwriter as an argument so in subclasses you'd have to call `doWithIndexWriter` if you want to commit something. I don't think I'm actually using commit in subclasses so I could remove it now.

**Pull request merged**